### PR TITLE
Enhancement: API Keys Management

### PR DESCRIPTION
**Pull request by aybruhm**
The following have been worked on:

- setup redis service on the django test workflow
- removed default throttle rates
- created admin client, admin client API key and override client API key admin add, change and delete permission
- created client and client API key models
- modified get_user_apikey method in the test_views
- override the API key throttling class to support custom scope and rate
- override the default dispatch method on API views to pass the request to the throttling class

**EDIT:** modified the throttle test

### CODE REVIEW

**Comment by aybruhm** [comment]
To create an api key, follow the below process:

1. Create a Client
```python
from academia.models import Client

client = Client.objects.create(
    name="abram-client", 
    email="abram-client@email.com", 
    is_verified=True, 
    client_type=Client.ClientTypes.DEVELOPER
)
client.save()
```

2. Create a Client APIKey
```python
from datetime import datetime import timedelta
from academia.models import ClientAPIKey

_, api_key = ClientAPIKey.objects.create_key(
    name="abram-client-apikey", 
    expiry_time=ClientAPIKey.ExpireWhen.THIRTY_DAYS,

    # replace 30 with the value of expiry_time -> could be 30, 60 or 90
    expiry_date=datetime.now() + timedelta(days=int("30")),

    client=client,
    scope="abram",
    rate=5000,
)

```

`_` is the client_apikey object and `api_key` is the client's API key. Feel free to let me know if you need clarification on something.

File: .github/workflows/django.yml
Comment (devgenix) [inline .github/workflows/django.yml:1]: Beautiful work here, thank you for this.

File: academia/admin.py [hunk 1]
```diff
@@ -1,47 +1,131 @@
...
+    list_filter = ("name", "code")
+    empty_value_display = "-empty field-"
+
+
+class AdminClient(admin.ModelAdmin):
+    list_display: list = ["id", "name", "email", "is_verified", "client_type"]
+    list_filter: list = ["is_verified", "client_type"]
+
+
+class AdminClientAPIKey(admin.ModelAdmin):
```
Comment (devgenix) [inline academia/admin.py:97]: Although I see it structured here, I couldn't create a ClientAPIKey in the Django Admin.

File: academia/models.py [hunk 2]
```diff
@@ -7,6 +7,75 @@
...
+    """
+
+    class ClientTypes(models.Choices):
+        DEVELOPER = "developer"
+        ORGANISATION = "organisation"
+
+    id = models.UUIDField(default=uuid.uuid4, primary_key=True, editable=False)
+    name = models.CharField(max_length=255, null=True, blank=True)
+    email = models.EmailField(max_length=255, unique=True)
+    is_verified = models.BooleanField(default=False)
```
Comment (devgenix) [inline academia/models.py:34]: How is verification done?

**Review by devgenix — CHANGES_REQUESTED** [review]

**Comment by devgenix** [comment]
> To create an api key, follow the below process:
> 
>     1. Create a Client
> 
> 
> ```python
> from academia.models import Client
> 
> client = Client.objects.create(
>     name="abram-client", 
>     email="abram-client@email.com", 
>     is_verified=True, 
>     client_type=Client.ClientTypes.DEVELOPER
> )
> client.save()
> ```
> 
>     2. Create a Client APIKey
> 
> 
> ```python
> from datetime import datetime import timedelta
> from academia.models import ClientAPIKey
> 
> _, api_key = ClientAPIKey.objects.create_key(
>     name="abram-client-apikey", 
>     expiry_time=ClientAPIKey.ExpireWhen.THIRTY_DAYS,
> 
>     # replace 30 with the value of expiry_time -> could be 30, 60 or 90
>     expiry_date=datetime.now() + timedelta(days=int("30")),
> 
>     client=client,
>     scope="abram",
>     rate=5000,
> )
> ```
> 
> `_` is the client_apikey object and `api_key` is the client's API key. Feel free to let me know if you need clarification on something.

I understand but does this have to be done programmatically?

**Pull request merged**